fiat-roles/src/main/java/com/netflix/spinnaker/fiat/permissions/RedisPermissionsRepository.java

```diff
@@ -206,40 +288,36 @@ public RedisPermissionsRepository put(@NonNull UserPermission permission) {
 
   private Optional<UserPermission> getFromRedis(@NonNull String id) {
     try {
+      TimeoutContext timeoutContext =
+          new TimeoutContext(
+              String.format("getPermission for user: %s", id),
+              clock,
+              configProps.getRepository().getGetPermissionTimeout());
       boolean userExists =
-          UNRESTRICTED.equals(id)
-              || redisClientDelegate.withCommandsClient(
-                  c -> {
-                    return c.sismember(allUsersKey(), id);
-                  });
+          UNRESTRICTED.equals(id) || redisRead(timeoutContext, c -> c.sismember(allUsersKey(), id));
       if (!userExists) {
         return Optional.empty();
       }
-      final RawUserPermission userResponseMap = new RawUserPermission();
-      redisClientDelegate.withMultiKeyPipeline(
-          p -> {
-            resources.stream()
-                .map(Resource::getResourceType)
-                .forEach(
-                    r -> {
-                      String userKey = userKey(id, r);
-                      Response<Map<String, String>> resourceMap = p.hgetAll(userKey);
-                      userResponseMap.put(r, resourceMap);
-                    });
-            Response<Boolean> admin = p.sismember(adminKey(), id);
-            p.sync();
-            userResponseMap.isAdmin = admin.get();
+      UserPermission userPermission = new UserPermission().setId(id);
+      resources.forEach(
+          r -> {
+            ResourceType resourceType = r.getResourceType();
+            String userKey = userKey(id, resourceType);
+            Map<String, String> resourcePermissions = hgetall(timeoutContext, userKey);
+            userPermission.addResources(extractResources(resourceType, resourcePermissions));
           });
-
-      UserPermission userPermission = getUserPermission(id, userResponseMap);
+      userPermission.setAdmin(redisRead(timeoutContext, c -> c.sismember(adminKey(), id)));
       if (!UNRESTRICTED.equals(id)) {
         userPermission.merge(getUnrestrictedUserPermission());
       }
       return Optional.of(userPermission);
+    } catch (SpinnakerException se) {
+      throw se;
     } catch (Exception e) {
-      log.error("Storage exception reading " + id + " entry.", e);
+      String message = String.format("Storage exception reading %s entry.", id);
+      log.error(message, e);
+      throw new PermissionReadException(message, e);
     }
-    return Optional.empty();
```

Comment:
no longer masking exceptions and treating them as a user not found

I'm reasonably happy with introducing more error responses from the API in this change because I think it is more correct / less surprising than just failing with a not found and there are retries on all the read operations so the case where the not-found would kick into our fallback in authorize controller as a fallback for a redis exception (rather than just as a fallback for the user not having transited gate) I think is better handled in here with retrying the read operations against redis

Reply:
Agreed.  This is way cleaner and better.  Nice improvement.